raft/src/test/java/org/apache/kafka/raft/MockLog.java

```diff
@@ -198,19 +238,51 @@ public Records read(long startOffset, OptionalLong endOffset) {
     @Override
     public void assignEpochStartOffset(int epoch, long startOffset) {
         if (startOffset != endOffset())
-            throw new IllegalStateException("Can only assign epoch for the end offset");
+            throw new IllegalArgumentException(
+                "Can only assign epoch for the end offset " + endOffset() + ", but get offset " + startOffset);
         epochStartOffsets.add(new EpochStartOffset(epoch, startOffset));
     }
 
-    public static class LogEntry {
+    static class LogEntry {
         final long offset;
         final int epoch;
         final SimpleRecord record;
+        final ControlRecordType controlRecordType;
 
-        private LogEntry(long offset, int epoch, SimpleRecord record) {
+        private LogEntry(long offset,
+                         int epoch,
+                         SimpleRecord record,
+                         ControlRecordType controlRecordType) {
             this.offset = offset;
             this.epoch = epoch;
             this.record = record;
+            this.controlRecordType = controlRecordType;
+        }
+
+        static LogEntry with(long offset, int epoch, SimpleRecord record) {
+            return new LogEntry(offset, epoch, record, ControlRecordType.UNKNOWN);
```

Comment:
nit: I think this is sort of an abuse of the `UNKNOWN` type. This value is reserved when the record is a control record, but it uses a type that we do not know about. Instead, maybe we could use `Optional<ControlRecordType>`?

Reply:
Sounds good